### DISCUSSION TITLE
PyEncoder using self.im.tobytes()

### DISCUSSION THREAD

**Discussion by duaneg3** [python-pillow/Pillow#7551] [Q&A] [answered]
I want to have a new encoder which handles ZLIB.  The easiest way to do this is to generate a raw stream first and then encode it.  When I try to use tobytes() in the encoder it complains:
```python
       def encode(self, bufsize):
           buf1 = self.im.tobytes("raw")
```
```
     AttributeError: 'ImagingCore' object has no attribute 'tobytes'
```
Is there anyway to get this to work??

  The work-around seems to be to write a custom _save routine  
```python
def _save(im, fp, filename):
    buf1 = im.tobytes()
    buf2 = zlib.compress(buf1);
    fp.write(buf2)
```
so then can save to an io.BytesIO
```python
        data2 = io.BytesIO()
        img.save(data2,"ZLIB")
```
However, given I never want to save to a file, this does not seem as clear ..

**Comment by radarhere** [comment, marked as answer]
> buf1 = self.im.tobytes("raw")

The simplest way to do this would be to construct an image around the ImagingCore object.
```python
buf1 = Image.Image()._new(self.im).tobytes("raw")
```

If you would like to work at a lower level, you could start extracting the code from [`tobytes()`](https://github.com/python-pillow/Pillow/blob/902055fbfe559e0d13960c26f4ed6251ba0faaf4/src/PIL/Image.py#L708-L758) instead.
```python
encoder = Image.core.raw_encoder(self.im.mode, self.im.mode)
encoder.setimage(self.im)
bufsize = max(65536, self.im.size[0] * 4)
bytes_consumed, errcode, data = encoder.encode(bufsize)
if errcode < 0:
	msg = f"encoder error {errcode} in tobytes"
	raise RuntimeError(msg)
buf1 = data
```

**Reply by duaneg3** [reply]
This worked very nicely!!  Can make a simple ZLIB encoder this way:
```python
class ZLIBEncoder(ImageFile.PyEncoder):
    _pushes_fd = True 
    def encode(self, bufsize):
        buf1 = Image.Image()._new(self.im).tobytes("raw")
        buf2 = zlib.compress(buf1)
        return len(buf2), 1, buf2
Image.register_encoder("ZLIB", ZLIBEncoder)
```
Can also use the same method to create a JPEG encoder:
```python
class JPEGEncoder(ImageFile.PyEncoder):
    _pushes_fd = True
    def encode(self, bufsize):
        data2 = io.BytesIO()
        Image.Image()._new(self.im).save(data2,'JPEG')
        buf1 = b""+data2.getvalue()
        return len(buf1), 1, buf1
Image.register_encoder("JPEG", JPEGEncoder)
```
